modalsheet/src/main/java/eu/wewox/modalsheet/ModalSheet.kt

```diff
@@ -96,259 +79,134 @@ public fun <T> ModalSheet(
+ * bottom sheet is visible. If the color passed is [Color.Unspecified], then a scrim will no
+ * longer be applied and the bottom sheet will not block interaction with the rest of the screen
+ * when visible.
  * @param content The content of the bottom sheet.
  */
 @ExperimentalSheetApi
 @Composable
 public fun ModalSheet(
     visible: Boolean,
     onDismiss: () -> Unit,
-    shape: Shape = MaterialTheme.shapes.large.copy(
-        bottomEnd = CornerSize(0),
-        bottomStart = CornerSize(0)
-    ),
-    backgroundColor: Color = MaterialTheme.colors.surface,
-    swipeEnabled: Boolean = true,
-    scrimClickEnabled: Boolean = true,
-    scrim: @Composable BoxScope.(Boolean) -> Unit = {
-        Scrim(it, onScrimClick = if (scrimClickEnabled) onDismiss else NoOpLambda)
-    },
-    content: @Composable ModalSheetScope.() -> Unit,
+    cancelable: Boolean = true,
+    shape: Shape = ModalSheetDefaults.shape,
+    elevation: Dp = ModalSheetDefaults.elevation,
+    backgroundColor: Color = ModalSheetDefaults.backgroundColor,
+    contentColor: Color = contentColorFor(backgroundColor),
+    scrimColor: Color = ModalSheetDefaults.scrimColor,
+    content: @Composable ColumnScope.() -> Unit,
 ) {
-    // If content animated by AnimatedVisibility in PopupBody is still visible (in composition)
-    var contentVisible by remember { mutableStateOf(false) }
-
-    if (visible || contentVisible) {
-        FullscreenPopup(
-            onDismiss = onDismiss
-        ) {
-            PopupBody(
-                visible = visible,
-                scrim = scrim,
-                content = {
-                    // Set contentVisible true when the composable enters composition and false when it leaves
-                    DisposableEffect(Unit) {
-                        contentVisible = true
-                        onDispose { contentVisible = false }
-                    }
-
-                    val scope = remember { ModalSheetScopeImpl() }
-                    SwipeableContent(
-                        swipeEnabled = swipeEnabled,
-                        onSwipedAway = onDismiss,
-                        scrollState = scope.scrollState
-                    ) {
-                        Surface(
-                            shape = shape,
-                            color = backgroundColor,
-                            modifier = Modifier.fillMaxWidth()
-                        ) {
-                            scope.content()
-                        }
-                    }
-                }
-            )
+    val sheetState = rememberModalBottomSheetState(
+        skipHalfExpanded = true,
+        initialValue = ModalBottomSheetValue.Hidden,
+        confirmStateChange = {
+            // Intercept and disallow hide gesture / action
+            if (it == ModalBottomSheetValue.Hidden && cancelable) {
+                onDismiss()
+            }
+            false
```

Comment:
If we can't disable the swipe gesture, is there really a point in having this cancelable flag instead of exposing the underlying ModalBottomSheetState? If we want to give the cleaner dialog-like API, maybe we should also have a public overload with the ModalBottomSheetState exposed so we can cover even the more complex use cases.
What do you think?

Reply:
> If we can't disable the swipe gesture, is there really a point in having this cancelable flag instead of exposing the underlying ModalBottomSheetState?

I leave this flag in `ModalSheets` overloads which do not use sheet state directly. I find those overloads are simpler to use, so as for me it is nice to have such flag in their APIs.

> If we want to give the cleaner dialog-like API, maybe we should also have a public overload with the ModalBottomSheetState exposed so we can cover even the more complex use cases

Sure, I have added `ModalSheet` with sheet state to the public API and have added an example usage.

Also in last commit I have addressed de-sync issues we discussed on Slack, however the API needs to be changed. There is no more `onDismiss`, but `onVisibleChange` and `onDataChanged` lambdas in "higher-level" overloads. 

Most de-sync issues were due to the interaction collision between `hide()` / `show()` and user's input. When user "interacts" with bottom sheet each `hide()` / `show()` are getting cancelled due to higher priority of user input, see `MutatePriority` & `DraggableState::drag` if you are interested.

There were a lot of trials and errors to fix de-sync and this solution looks like the best I can get 😅